pkg/client/record/event_test.go

```diff
@@ -56,6 +61,57 @@ func (t *testEventSink) Update(e *api.Event) (*api.Event, error) {
 	return e, nil
 }
 
+// PatchEvent records the event for testing.
+func (t *testEventSink) Patch(e *api.Event, p []byte) (*api.Event, error) {
+	if t.OnPatch != nil {
+		return t.OnPatch(e, p)
+	}
+	return e, nil
+}
+
+type OnCreateFunc func(*api.Event) (*api.Event, error)
+
+func OnCreateFactory(testCache map[string]*api.Event, createEvent chan *api.Event) OnCreateFunc {
+	return func(event *api.Event) (*api.Event, error) {
+		// because the accuracy of time decreases to second-level after time is
+		// encoded, we have to manually shift the timestamp here, so the
+		// timestamp in patch will be different even if the same event occurs in
+		// rapid succession.
+		shift := time.Duration(-10 * time.Second)
+		event.FirstTimestamp = unversioned.NewTime(event.FirstTimestamp.Time.Add(shift))
```

Comment:
> 
```
func OnPatchFactory(testCache map[string]*api.Event, createEvent chan<- *api.Event, ticker func() time.Time) OnPatchFunc {
    return func(event *api.Event) (*api.Event, error) {
        ts := ticker()
        // Modify timestamps appropriately
        ...
    }
}
```

Introducing a fake clock would be a good idea, but we shouldn't manually change the patch here, the point of this unit test is verifying the generated patch is correct.
We just need a difference between the time when an event is created and when it's patched. The introduced 10 seconds shift will eliminate flake.